performance-tests/bench_2/common/util.cpp

```diff
@@ -137,4 +140,327 @@ uint32_t one_at_a_time_hash(const uint8_t* key, size_t length)
   return hash;
 }
 
+void update_stats_for_tags(std::unordered_map<std::string, uint64_t>& stats,
+  const Builder::StringSeq& reported_tags,
+  const std::unordered_set<std::string>& input_tags,
+  const Builder::ConstPropertyIndex& prop)
+{
+  for (CORBA::ULong i = 0; i < reported_tags.length(); ++i) {
+    const std::string tag(reported_tags[i]);
+    if (input_tags.find(tag) != input_tags.end()) {
+      if (stats.find(tag) == stats.end()) {
+        stats[tag] = prop->value.ull_prop();
+      }
+      else {
+        stats[tag] += prop->value.ull_prop();
+      }
+    }
+  }
+}
+
+void update_details_for_tags(std::unordered_map<std::string, std::string>& details,
+  const Builder::StringSeq& reported_tags,
+  const std::unordered_set<std::string>& input_tags,
+  const std::string& detail)
+{
+  for (CORBA::ULong i = 0; i < reported_tags.length(); ++i) {
+    const std::string tag(reported_tags[i]);
+    if (input_tags.find(tag) != input_tags.end()) {
+      details[tag] += detail;
+    }
+  }
+}
+
+void consolidate_tagged_stats(std::unordered_map<std::string, Bench::SimpleStatBlock>& stats,
+  const Builder::StringSeq& reported_tags,
+  const std::unordered_set<std::string>& input_tags,
+  const Bench::ConstPropertyStatBlock& data)
+{
+  for (CORBA::ULong i = 0; i < reported_tags.length(); ++i) {
+    const std::string tag(reported_tags[i]);
+    if (input_tags.find(tag) != input_tags.end()) {
+      stats[tag] = consolidate(stats[tag], data.to_simple_stat_block());
+    }
+  }
+}
+
+int handle_report(const Bench::TestController::Report& report,
```

Comment:
The behavior of handle_reports is unique to the test_controller, as it's writing out results to the screen and returning with an exit code. I don't think this is generic enough to move over.  Instead, I was thinking parts of this function, specifically the parts that loop over the parts of the report, could be factored out into common code for use in a scraping tool. Looking at this now, I'm not sure it will make sense for test_controller to use whatever we factor out since it will be faster to collect all the stats at once if we're interested in all the stats.

But for the scraping tool, I think we may sometimes be interested in one stat at a time, so we could have functions that, for example, loop over all datareader & datawriter reports and consolidates stat blocks of certain names (e.g. "discovery_delta" and "round_trip_latency") that match a tagged expression (e.g. "'signal' AND NOT 'best effort' or 'any'). Or, as another example, we could have a function that loops over all node controller reports and collects stat blocks of a certain name (e.g. "cpu_percent") and returns the individual blocks in a sequence with associated node names, so we can plot them individually.

All that to say, I think we'd better move this back into test_controller until we figure out what our common collection functions need to look like. Otherwise, the rest of this looks fine.

Reply:
By "the rest of this looks file", I think you're saying we should move update_stats_for_tags, update_details_for_tags and consolidate_tagged_stats into util, but keep handle_report in main.  Is that correct?